iocore/net/quic/QUICStreamState.cc

```diff
@@ -24,185 +24,92 @@
-      return true;
-    }
-    break;
-  case State::DataRead:
-    if (type == QUICFrameType::STOP_SENDING) {
-      return true;
-    }
-    break;
-  case State::ResetRecvd:
-    // It should not send any frame after receiving RESET_STREAM
-    break;
-  case State::ResetRead:
-    // It should not send any frame after receiving RESET_STREAM
-    break;
-  default:
-    ink_assert(!"Unknown state");
-    break;
+  if (type != QUICFrameType::STOP_SENDING && type != QUICFrameType::MAX_STREAM_DATA) {
+    return false;
+  }
+
+  QUICReceiveStreamState state = this->get();
+  // The receiver only sends MAX_STREAM_DATA in the “Recv” state.
+  if (type == QUICFrameType::MAX_STREAM_DATA && state == QUICReceiveStreamState::Recv) {
+    return true;
+  }
+
+  // A receiver can send STOP_SENDING in any state where it has not received a RESET_STREAM frame; that is states other than “Reset
+  // Recvd” or “Reset Read”.
+  if (type == QUICFrameType::STOP_SENDING && state != QUICReceiveStreamState::ResetRecvd &&
+      state != QUICReceiveStreamState::ResetRead) {
+    return true;
   }
 
   return false;
 }
 
 bool
-QUICReceiveStreamState::is_allowed_to_receive(const QUICFrame &frame) const
+QUICReceiveStreamStateMachine::is_allowed_to_receive(const QUICFrame &frame) const
 {
   return this->is_allowed_to_receive(frame.type());
 }
 
 bool
-QUICReceiveStreamState::is_allowed_to_receive(QUICFrameType type) const
+QUICReceiveStreamStateMachine::is_allowed_to_receive(QUICFrameType type) const
 {
-  // Return true or break out the switch to return false
-  switch (this->get()) {
-  case State::_Init:
-    if (type == QUICFrameType::STREAM || type == QUICFrameType::RESET_STREAM || type == QUICFrameType::MAX_STREAM_DATA ||
-        type == QUICFrameType::STREAM_DATA_BLOCKED) {
-      return true;
-    }
-    break;
-  case State::_Invalid:
-    // Everthing is invalid on this state
-    break;
-  case State::Recv:
-    return true;
-  case State::SizeKnown:
-    return true;
-  case State::DataRecvd:
+  // always allow receive these frames.
+  if (type == QUICFrameType::STREAM || type == QUICFrameType::STREAM_DATA_BLOCKED || type == QUICFrameType::RESET_STREAM) {
     return true;
-  case State::DataRead:
-    return true;
-  case State::ResetRecvd:
-    return true;
-  case State::ResetRead:
-    return true;
-  default:
-    ink_assert(!"Unknown state");
-    break;
   }
 
   return false;
 }
 
 void
-QUICReceiveStreamState::update_with_sending_frame(const QUICFrame &frame)
+QUICReceiveStreamStateMachine::update_with_sending_frame(const QUICFrame &frame)
 {
 }
 
 void
-QUICReceiveStreamState::update_with_receiving_frame(const QUICFrame &frame)
+QUICReceiveStreamStateMachine::update_with_receiving_frame(const QUICFrame &frame)
 {
+  // The receiving part of a stream initiated by a peer (types 1 and 3 for a client, or 0 and 2 for a server) is created when the
```

Comment:
[nitpicking] It'd better to avoid using underlying type directly in comments.

Reply:
This comment are from specs without any changes.

Reply:
Oh, I found it on "3.2.  Receiving Stream States". Never mind.